pkg/reconciler/cache/replication/replication_reconcile.go

```diff
@@ -24,194 +24,178 @@ import (
 // the replication function handles the following cases:
-//  1. creation of the object in the cache server when the cached object is not found by retrieveLocalObject
-//  2. deletion of the object from the cache server when the original/local object was removed OR was not found by retrieveLocalObject
+//  1. creation of the object in the cache server when the cached object is not found by getLocalCopy
+//  2. deletion of the object from the cache server when the original/local object was removed OR was not found by getLocalCopy
 //  3. modification of the cached object to match the original one when meta.annotations, meta.labels, spec or status are different
-func (c *controller) reconcileObject(ctx context.Context,
-	key string, gvr schema.GroupVersionResource, gvk schema.GroupVersionKind,
-	retrieveCacheObject func(gvr schema.GroupVersionResource, cluster logicalcluster.Name, namespace, name string) (interface{}, error),
-	retrieveLocalObject func(cluster logicalcluster.Name, namespace, name string) (interface{}, bool, error)) error {
-	cluster, namespace, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+func (r *reconciler) reconcile(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx).WithValues("reconcilerKey", key)
+
+	clusterName, ns, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
 	if err != nil {
-		return err
+		runtime.HandleError(err)
+		return nil
 	}
-	cacheObject, err := retrieveCacheObject(gvr, cluster, namespace, name)
+
+	localCopy, err := r.getLocalCopy(clusterName, ns, name)
 	if err != nil && !errors.IsNotFound(err) {
-		return err
+		runtime.HandleError(err)
+		return nil
 	}
-	localObject, replicate, err := retrieveLocalObject(cluster, namespace, name)
+	localExists := !errors.IsNotFound(err)
+
+	globalCopy, err := r.getGlobalCopy(clusterName, ns, name)
 	if err != nil && !errors.IsNotFound(err) {
-		return err
-	}
-	if errors.IsNotFound(err) {
-		// issue a live GET to make sure the localObject was removed
-		_, err = c.dynamicKcpLocalClient.Cluster(cluster.Path()).Resource(gvr).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
-		if err == nil {
-			return fmt.Errorf("the informer used by this controller is stale, the following %s resource was found on the local server: %s/%s/%s but was missing from the informer", gvr, cluster, namespace, name)
-		}
-		if !errors.IsNotFound(err) {
-			return err
-		}
-	} else if !replicate {
-		localObject = nil // like if it wasn't there
+		runtime.HandleError(err)
+		return nil
 	}
+	globalExists := !errors.IsNotFound(err)
 
-	var unstructuredCacheObject *unstructured.Unstructured
-	var unstructuredLocalObject *unstructured.Unstructured
-	if cacheObject != nil {
-		unstructuredCacheObject, err = toUnstructured(cacheObject)
-		if err != nil {
-			return err
+	// local is gone or being deleted. Delete in cache.
+	if !localExists || !localCopy.GetDeletionTimestamp().IsZero() {
```

Comment:
How does it play with finalizers? Doesn't it remove the object from the cache when finalizers may be holding the local copy? Do we want that?

Reply:
we patched the cache server to ignore `Finalizers`

Reply:
very good question. So we should keep updating it, but filter the deletion timestsamp?

Does the cache-server ignore or follow finalizers?

Reply:
removing an object is enough

Reply:
follow-up